plugins/googlesignin/class.googlesignin.plugin.php

```diff
@@ -0,0 +1,124 @@
+     */
+    public function __construct() {
+        $this->setProviderKey('googlesignin');
+        $this->settingsView = 'plugins/settings/googlesignin';
+    }
+
+    /**
+     *  Return all the information saved in provider table, add hardcoded values.
+     *
+     * @return array Stored provider data (secret, client_id, etc.).
+     */
+    public function provider() {
+        if (!$this->provider) {
+            $this->provider = Gdn_AuthenticationProviderModel::getProviderByKey($this->providerKey);
+
+            // These URLs are added here instead of being stored in the DB in case they ever change, we will not have to update the DB.
+            $this->provider['AuthorizeUrl'] = self::AUTHORIZEURL;
+            $this->provider['TokenUrl'] = self::TOKENURL;
+            $this->provider['ProfileUrl'] = self::PROFILEURL;
+
+            // Scope
+            $this->provider['AcceptedScope'] = self::ACCEPTEDSCOPE;
+
+            // Translate claims coming back from Google.
+            $this->provider['ProfileKeyName'] = self::PROFILENAME;
+            $this->provider['AcceptedScope'] = 'email openid profile';
+
+            $this->provider['ProfileKeyUniqueID'] = 'sub';
+            $this->provider['ProfileKeyFullName'] = null;
+
+            // provider Name is what puts the text on the button and determines the CSS class.
+            $this->provider['Name'] = 'Google';
+        }
+
+        return $this->provider;
+    }
+
+    /**
+     * Create a controller to deal with plugin settings in dashboard.
+     *
+     * @param Gdn_Controller $sender.
+     * @param Gdn_Controller $args.
+     */
+    public function settingsEndpoint($sender, $args) {
+        $sender->permission('Garden.Settings.Manage');
+        $model = new Gdn_AuthenticationProviderModel();
+
+        /* @var Gdn_Form $form */
+        $form = new Gdn_Form();
+        $form->setModel($model);
+        $sender->Form = $form;
+
+        if (!$form->authenticatedPostBack()) {
+            $provider = $this->provider();
+            $form->setData($provider);
+        } else {
+
+            $form->setFormValue('AuthenticationKey', $this->getProviderKey());
+
+            $sender->Form->validateRule('AssociationKey', 'ValidateRequired', 'You must provide a unique AccountID.');
+            $sender->Form->validateRule('AssociationSecret', 'ValidateRequired', 'You must provide a Secret');
+
+            // To satisfy the AuthenticationProviderModel, create a BaseUrl.
+            $baseUrlParts = parse_url($form->getValue('AuthorizeUrl'));
+            $baseUrl = (val('scheme', $baseUrlParts) && val('host', $baseUrlParts)) ? val('scheme', $baseUrlParts).'://'.val('host', $baseUrlParts) : null;
+            if ($baseUrl) {
+                $form->setFormValue('BaseUrl', $baseUrl);
+                $form->setFormValue('SignInUrl', $baseUrl); // kludge for default provider
+            }
+            if ($form->save()) {
+                $sender->informMessage(t('Saved'));
```

Comment:
Do we translate interface elements and messages for dashboard generally?
If **yes**, all message strings in this plugin need to be wrapped with `t()` function.
Or if **no**, then we can remove `t()` call here.